tests/sub_group/sub_group_api.cpp

```diff
@@ -0,0 +1,430 @@
+    const size_t sub_group_count =
+        results[get_idx(i, member_function::group_range)];
+    REQUIRE(wg_id < work_group_count);  // bounds check
+    if (sub_group_counts[wg_id] == empty) {
+      sub_group_counts[wg_id] = sub_group_count;
+    } else {
+      // if a previous work-item from the same work-group already set the value,
+      // check that it is equal
+      REQUIRE((sub_group_count == sub_group_counts[wg_id]));
+    }
+  }
+  // check that all elements were assigned a value, meaning that
+  // all unique work-groups are processed
+  REQUIRE(std::all_of(sub_group_counts.begin(), sub_group_counts.end(),
+                      [=](size_t count) { return count != empty; }));
+
+  // Check that each work-group has as many sub-groups as reported previously.
+  // ---------------------------------------------------------------------------
+
+  // for each work-group, the number of sub-groups before it
+  std::vector<size_t> work_group_offsets(work_group_count);
+  std::exclusive_scan(sub_group_counts.begin(), sub_group_counts.end(),
+                      work_group_offsets.begin(), 0);
+
+  // the number of sub-groups across all work-groups
+  size_t sub_group_count =
+      std::reduce(sub_group_counts.begin(), sub_group_counts.end(), 0);
+  // for each sub-group, whether there exists at least one local item with that
+  // sub-group id
+  std::vector<bool> sub_group_index_seen(sub_group_count, false);
+  // populate sub_group_index_seen
+  for (size_t i = 0; i < linear_execution_range; i++) {
+    const size_t wg_id =
+        results[get_idx(i, member_function::group_group_linear_id)];
+    REQUIRE(wg_id < work_group_count);  // bounds check
+    const size_t local_sg_id = results[get_idx(i, member_function::group_id)];
+    // #sub-groups before this work-group + sub-group index in this work-group
+    //  = global sub-group index
+    const size_t global_sg_id = work_group_offsets[wg_id] + local_sg_id;
+    REQUIRE(global_sg_id < sub_group_count);  // bounds check
+    sub_group_index_seen[global_sg_id] = true;
+  }
+  CHECK(std::all_of(sub_group_index_seen.begin(), sub_group_index_seen.end(),
+                    [](bool index_is_seen) { return index_is_seen; }));
+
+  // Check that all work-items in the same sub-group report the same sub-group
+  // size. Use REQUIRE as sub_group_sizes is used by next checks.
+  // ---------------------------------------------------------------------------
+
+  // for each sub-group, the number of work-items it contains
+  std::vector<size_t> sub_group_sizes(sub_group_count, empty);
+  // populate and check sub_group_sizes by iterating over the work-items
+  for (size_t i = 0; i < linear_execution_range; i++) {
+    const size_t wg_id =
+        results[get_idx(i, member_function::group_group_linear_id)];
+    REQUIRE(wg_id < work_group_count);  // bounds check
+    const size_t local_sg_id = results[get_idx(i, member_function::group_id)];
```

Comment:
You could simplify the code everywhere with
```suggestion
    const size_t local_sg_id = result(i, member_function::group_id);
```
by defining the right `result` function or lambda.

Reply:
Nice suggestion, I've applied it.